### PR TITLE
Aimbot Hitchance

### PR DESCRIPTION
**Pull request by dapaster**
For those who had trouble adding this

### CODE REVIEW

**Comment by aljawary** [comment]
i got 18 errors in angle.h
https://prnt.sc/sw8v3g

**Comment by dapaster** [comment]
@aljawary this is for all weapon

**Comment by aljawary** [comment]
my bad i forgot to check auto shot on

**Comment by hastydev** [comment]
why did you delete inaccuracy? that's auto shoot thing

**Comment by notgoodusename** [comment]
> why did you delete inaccuracy? that's auto shoot thing

hitchance replaces that

**Comment by Cyk-Fad** [comment]
niiiiiiiiiiiiiiiiiiiiiiiiice

**Comment by zxsleebu** [comment]
Aimbot with hc 100 can't hit ppl while they jumping

**Comment by notgoodusename** [comment]
> Aimbot with hc 100 can't hit ppl while they jumping

using 100 hitchance wont shoot most of the time, use less than 80

**Comment by zxsleebu** [comment]
> > Aimbot with hc 100 can't hit ppl while they jumping
> 
> using 100 hitchance wont shoot most of the time, use less than 80

thanks, i know, but it will not hit ppl with other REALLY HVH cheats. so i decided to set hc to 100. but sometimes at small distance it doesn't hit people without aa.

File: Osiris/SDK/Angle.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+public:
+    Vector forward{ };
+    Vector right{ };
+    Vector up{ };
+
+    Angle(const Vector& v) noexcept :sp{ sinf(degreesToRadians(v.x)) }, cp{ cosf(degreesToRadians(v.x)) }, sy{ sinf(degreesToRadians(v.y)) }, cy{ cosf(degreesToRadians(v.y)) }, sr{ sinf(degreesToRadians(v.z)) }, cr{ cosf(degreesToRadians(v.z)) },
+        forward{ get(0) }, right{ get(1) }, up{ get(2) } { }
+};
```
Comment (lofcz) [inline Osiris/SDK/Angle.h:37]: did you mean this?  

```cpp
#pragma once
#include "SDK/Utils.h"
#include "SDK/Vector.h"

class Angle
{
    float sp{ 0.0f }, cp{ 0.0f };
    float sy{ 0.0f }, cy{ 0.0f };
    float sr{ 0.0f }, cr{ 0.0f };

    constexpr Vector get(unsigned int i) noexcept
    {
        switch (i)
        {
        case 0:
            return { (cp * cy), (cp * sy), (-sp) };
            break;
        case 1:
            return { -1 * sr * sp * cy + -1 * cr * -sy, -1 * sr * sp * sy + -1 * cr * cy, -1 * sr * cp };
            break;
        case 2:
            return { cr * sp * cy + -sr * -sy, cr * sp * sy + -sr * cy, cr * cp };
            break;
        default:
            return { };
            break;
        }
    }

public:
    Vector forward{ };
    Vector right{ };
    Vector up{ };

    Angle(const Vector& v) noexcept :sp{ sinf(degreesToRadians(v.x)) }, cp{ cosf(degreesToRadians(v.x)) }, sy{ sinf(degreesToRadians(v.y)) }, cy{ cosf(degreesToRadians(v.y)) }, sr{ sinf(degreesToRadians(v.z)) }, cr{ cosf(degreesToRadians(v.z)) },
        forward{ get(0) }, right{ get(1) }, up{ get(2) } { }
};
```

**Comment by lofcz** [comment]
working fine, vs legiters hitchance about 30-50 is sufficient for rifles, 90+ with ssg/awp.

**Comment by kschou95** [comment]
i got 2 errors
"Error (active)	E0711	expression must have bool type (or be convertible to bool)"
and
"Error	C2451	conditional expression of type 'Vector' is illegal"

the "bestTarget" in if is highlighted
`if (bestTarget)
            {
                if (!hitChance(localPlayer.get(), entity, activeWeapon, bestAngle, cmd, config->aimbot[weaponIndex].hitchance))
                {
                    bestTarget = Vector{ };
                    continue;
                }
                break;
            }`

**Comment by dapaster** [comment]
> 
> 
> i got 2 errors
> "Error (active) E0711 expression must have bool type (or be convertible to bool)"
> and
> "Error C2451 conditional expression of type 'Vector' is illegal"
> 
> the "bestTarget" in if is highlighted
> `if (bestTarget) { if (!hitChance(localPlayer.get(), entity, activeWeapon, bestAngle, cmd, config->aimbot[weaponIndex].hitchance)) { bestTarget = Vector{ }; continue; } break; }`

replace bestTarget to bestTarget.notNull();

**Comment by spleatz** [comment]
Im crashing when im shooting, please help.

**Comment by zajkos** [comment]
update?

**Comment by thisisnotagithub** [comment]
Error	C2664	'bool canScan(Entity *,Entity *,const Vector &,const WeaponData *,int) noexcept': cannot convert argument 2 from 'Vector' to 'Entity *'

**Comment by SquidmarcHD** [comment]
i dont really want to necro this but i tried porting it to the newer version and it seems to be randomly shooting sometimes
afaik getspread is still 452 so idk why

**Comment by Shuvi-Moro** [comment]
I beg you, could you remake the code for the actual sources

**Comment by SquidmarcHD** [comment]
> I beg you, could you remake the code for the actual sources

rather use kittenpopos ghetto calculate hitchance 
`float spreadVal = (weapon->GetSpread() + weapon->GetInaccuracy());
float innaccuracyBase = (20 * (eyePos.Distance(shotTarget) / 130) * spreadVal);
float hitchance = std::clamp((1 / innaccuracyBase) * 175, 0.f, 100.f);`

**Pull request closed without merging**